### ISSUE TITLE
Not working with webpack's resolve.alias?

### ISSUE THREAD

**Issue by x1a0** [halt-hammerzeit/webpack-isomorphic-tools#22]
Basically I want to separate my themes (scss, icons, etc) out from my web app project. So each theme is a npm module under `node_modules`. In my component I have:

    import Styles from 'theme/sass/user.scss';

And in webpack config there is:

    resolve: {
        ...
        alias: { 'theme': 'theme-a' }
    }

Therefore I can easily change active theme by changing the alias here.

However since it's an isomorphic app, when it renders on server side it gives:

     Cannot find module 'theme/sass/user.scss'

If in component I hardcode it to `import Styles from 'theme-a/sass/user.scss';` and remove the resolve alias, everything works. (which means webpack-isomorphic-tools works well with `node_modules` I think).

**Comment by halt-hammerzeit** [comment]
Post your `webpack-stats.json` and `webpack-assets.json`

**Comment by x1a0** [comment]
###  webpack-assets.json

```
{
  "javascript": {
    "main": "/static/main-7fcd93e07365ad533827.js"
  },
  "styles": {},
  "assets": {
    "./src/app/styles/main.scss": {},
    "../theme-a/sass/login.scss": {
      "loginForm": "loginForm___faZqf"
    }
  }
}
```

`webpack-stats.json` is 2.6M with 28k lines... how shall I post it?

**Comment by halt-hammerzeit** [comment]
Is it from the case of
```
import Styles from 'theme/sass/user.scss';
```
and ` alias: { 'theme': 'theme-a' }` ?

post the stats through a gist

**Comment by x1a0** [comment]
Oh sorry it's from

    import Styles from 'theme/sass/login.scss

So `login.scss` instead of `user.scss`, and yes with that alias.

Here is `webpack-stats.json`: https://gist.github.com/x1a0/ba04ce3ff40bb548c4d3

**Comment by halt-hammerzeit** [comment]
So, is it correct that in case with `alias`ing you have these assets:

```
  "assets": {
    "./src/app/styles/main.scss": {},
    "../theme-a/sass/login.scss": {
      "loginForm": "loginForm___faZqf"
    }
  }
```
?

As I can see, the alias got in place.
But the path is weird: `../theme-a/sass/login.scss`
This path surely doesn't lead to `node_modules`

**Comment by halt-hammerzeit** [comment]
And webpack-stats clearly say that it's not in node_modules:

```
{
      "id": 84,
      "identifier": "/home/zhangxiao/projects/cth-poc/node_modules/css-loader/index.js?modules&importLoaders=2&sourceMap&localIdentName=[local]___[hash:base64:5]!/home/zhangxiao/projects/cth-poc/node_modules/autoprefixer-loader/index.js?browsers=last 2 version!/home/zhangxiao/projects/cth-poc/node_modules/sass-loader/index.js?outputStyle=expanded&sourceMap!/home/zhangxiao/projects/theme-a/sass/login.scss",
      "name": "./~/css-loader?modules&importLoaders=2&sourceMap&localIdentName=[local]___[hash:base64:5]!./~/autoprefixer-loader?browsers=last 2 version!./~/sass-loader?outputStyle=expanded&sourceMap!../theme-a/sass/login.scss",
  
```

See the ending of the `identifier`

**Comment by halt-hammerzeit** [comment]
As far as I understood, `alias` only works with `node_modules` folder.
Is that so?

**Comment by x1a0** [comment]
Or say `alias` only works with module resolving, in general yes only `node_modules`.

Actually as you said, I don't understand why `import Styles from 'theme-a/sass/login.scss'` works either... it generates exactly same `webpack-assets.json`:

```
{
  "javascript": {
    "main": "/static/main-3b07addf3624baab08fa.js"
  },
  "styles": {},
  "assets": {
    "./src/app/styles/main.scss": {},
    "../theme-a/sass/login.scss": {
      "loginForm": "loginForm___2m8hB"
    }
  }
}
```

**Comment by x1a0** [comment]
> And webpack-stats clearly say that it's not in node_modules:

This is because I use `npm link` now, so at the end of `identifier` it points to the actual path on my disk.

**Comment by halt-hammerzeit** [comment]
I can implement this `alias` feature but first it's required to have a clear understanding of how this `alias`ing works.
Webpack documentation is cryptic as usual.
If you're interested in me implementing this feature then you can write an explanation here (`alias` algorythm, how it works in Webpack).
[Two sentences in webpack docs](https://webpack.github.io/docs/resolving.html#aliasing) are definitely not specific enough.

**Comment by halt-hammerzeit** [comment]
> This is because I use npm link now, so at the end of identifier it points to the actual path on my disk.

Aha, so, your `npm link` leads outside of your project folder, I see.

So, suppose, we have `theme/...` require path.
It is resolved by Node.js internally to `theme` node_module which doesn't exist.
Actually, I think I can hook into Node.js module loading and intercept all `require()`s to module names from the `alias` section and substitute them.
I have all the information needed by far then.
I'll keep you posted when some progress is made.

**Comment by x1a0** [comment]
I have a rough idea. I guess it doesn't work because `webpack-isomophic-tools` sees the asset as 'theme-a/sass/login.scss' but in component it `require('theme/sass/login.scss')`.

Is that possible when hacking the `require()` method, read `resolve.alias` from webpack config, and apply the aliasing there too

**Comment by x1a0** [comment]
Just found another issue while trying: I moved the `npm link`ed module around to see whether the actual path matters and it stopped working.

in `webpack-assets.json`, the linked module has absolute path:

```

  "javascript": {
    "main": "/static/main-d4bdd95e2e22b16178b6.js"
  },
  "styles": {},
  "assets": {
    "./src/app/styles/main.scss": {},
    "/home/zhangxiao/test/theme-a/sass/login.scss": {
      "loginForm": "loginForm___3zhEz"
    }
  }
}
```

and `webpack-isomorphic-tools` gives error:

    [webpack-isomorphic-tools] [error] asset not found: ../../test/theme-a/sass/login.scss

**Comment by halt-hammerzeit** [comment]
Ooops, it closed it.
I've just released `v2.2.8`, try it and post back.

```
new tools(...).alias({ a: b }).server(.....)
```

**Comment by halt-hammerzeit** [comment]
As for your last issue, I didn't understand what you mean there.

**Comment by halt-hammerzeit** [comment]
I'll be likely here tomorrow.

**Comment by x1a0** [comment]
Hmm... still not working for me.

- It seems `2.2.8` is not built properly. `alias` is not a function of `tools`. However using source in master is fine.

- In my case I want to `import Styles from 'theme/sass/login.scss'`. I feel the `extension` resolver gets higher priority and run first so it does not use the `aliasing` resolver.

> As for your last issue, I didn't understand what you mean there.

In the assets file generated by `webpack-isomorphic-tools`, it uses absolute path: `"/home/zhangxiao/test/theme-a/sass/login.scss"`. However when `tools` tries to hack the server-side require, it looks up by relative path: `../../test/theme-a/sass/login.scss`. So it gives error that `asset not found`. Actually both paths point to the same file.

**Comment by halt-hammerzeit** [comment]
oops, I forgot to build it.
try v 2.2.9

**Comment by halt-hammerzeit** [comment]
If it won't work then post your assets, your stats and your debug logs.

**Comment by x1a0** [comment]
`webpack-assets.json`:

```
{
  "javascript": {
    "main": "/static/main-84cf2f71ca8ea931377a.js"
  },
  "styles": {},
  "assets": {
    "./src/app/styles/main.scss": {},
    "/home/zhangxiao/test/theme-a/sass/login.scss": {
      "loginForm": "loginForm___3zhEz"
    }
  }
}
```

`webpack-stats.json`: https://gist.github.com/890d17ff874e32b38be8

**Comment by halt-hammerzeit** [comment]
And the debug log?

**Comment by x1a0** [comment]
> And the debug log?

Hmm... how to generate that one...?

**Comment by halt-hammerzeit** [comment]
enable the { debug: true } flag in your webpack-isomorphic-tools configuration

**Comment by x1a0** [comment]
strange, it is `true` there. I can see debug info like:

```
[webpack-isomorphic-tools/plugin] [debug] running write assets webpack plugin
[webpack-isomorphic-tools/plugin] [debug]  (development mode is on)
[webpack-isomorphic-tools/plugin] [debug] writing webpack stats to /home/zhangxiao/projects/cth-poc/webpack-stats.json
[webpack-isomorphic-tools/plugin] [debug] getting javascript and styles for chunk "main"
[webpack-isomorphic-tools/plugin] [debug]  (got javascript)
[webpack-isomorphic-tools/plugin] [debug] No filter specified for "images" assets. Using a default one.
[webpack-isomorphic-tools/plugin] [debug] No path parser specified for "images" assets. Using a default one.
[webpack-isomorphic-tools/plugin] [debug] parsing assets of type "images"
[webpack-isomorphic-tools/plugin] [debug]  time taken: 1 ms
[webpack-isomorphic-tools/plugin] [debug] parsing assets of type "styleModules"
[webpack-isomorphic-tools/plugin] [trace] Adding assset "./src/app/styles/main.scss", module id 83 (in webpack-stats.debug.json)
[webpack-isomorphic-tools/plugin] [trace] Adding assset "/home/zhangxiao/test/theme-a/sass/login.scss", module id 84 (in webpack-stats.debug.json)
[webpack-isomorphic-tools/plugin] [debug]  time taken: 5 ms
[require-hook] [debug] Hooking into *.webpack-module files loading
[webpack-isomorphic-tools/plugin] [debug] compiling assets
[require-hook] [debug] Loading source code for /home/zhangxiao/projects/cth-poc/src/app/styles/main.scss.webpack-module
[require-hook] [debug] Loading source code for /home/zhangxiao/test/theme-a/sass/login.scss.webpack-module
[webpack-isomorphic-tools/plugin] [debug]  time taken: 1 ms
[webpack-isomorphic-tools/plugin] [debug] writing webpack assets info to /home/zhangxiao/projects/cth-poc/webpack-assets.json
webpack built 84cf2f71ca8ea931377a in 6291ms
```

But that file is not generated.

**Comment by halt-hammerzeit** [comment]
Yes, this is meant to be the debug log

**Comment by halt-hammerzeit** [comment]
But this is Webpack log.
And you should post here another log - your Node.js log

**Comment by x1a0** [comment]
I have `debug: true` in webpack-isomorphic-tools config. But it does not generate `webpack-stats.debug.json`.

**Comment by halt-hammerzeit** [comment]
`webpack-stats.debug.json` is now called `webpack-stats.json`.
Post here your Node.js log

**Comment by x1a0** [comment]
I hope this is what you need:

```
[webpack-isomorphic-tools] [debug] flushing require() caches
Error: Cannot find module 'theme/sass/login.scss'
    at Function.Module._resolveFilename (module.js:334:15)
    at Function.Module._load (module.js:284:25)
    at Module.require (module.js:363:17)
    at require (module.js:382:17)
    at Object.<anonymous> (/home/zhangxiao/projects/cth-poc/src/app/views/Login.jsx:7:16)
    at Module._compile (module.js:428:26)
    at normalLoader (/home/zhangxiao/projects/cth-poc/node_modules/babel-core/lib/api/register/node.js:199:5)
    at Object.require.extensions.(anonymous function) [as .jsx] (/home/zhangxiao/projects/cth-poc/node_modules/babel-core/lib/api/register/node.js:216:7)
    at Module.load (module.js:353:32)
    at Function.Module._load (module.js:308:12)
```

**Comment by halt-hammerzeit** [comment]
```
[webpack-isomorphic-tools] [debug] flushing require() caches
Error: Cannot find module 'theme/sass/login.scss'
```

Yes, this is it.
So, it tries to resolve 'theme/sass/login.scss' instead of 'theme-a/...'.
Post here your `webpack-isomorphic-tools` server instance code

**Comment by x1a0** [comment]
```
#!/usr/bin/env node

var fs = require('fs');
var path = require('path');
var babelrc = fs.readFileSync(path.resolve(__dirname, '..', '.babelrc'));

var config = {};

try {
  config = JSON.parse(babelrc);
} catch (err) {
  console.error('Failed to parsing .babelrc:');
  console.error(err);
}

require('babel/register')(config);

var WebpackIsomorphicTools = require('webpack-isomorphic-tools');

// server-side global constants
global.__PRODUCTION__ = process.env.NODE_ENV === 'production';
global.__DEVELOPMENT__ = !__PRODUCTION__;
global.__SERVER__ = true;
global.__CLIENT__ = false;

var webpackConfig = __PRODUCTION__ ? require('../webpack/prod.config') : require('../webpack/dev.config');

global.webpackIsomorphicToos = new WebpackIsomorphicTools(require('../webpack/isomorphic-tools-config'))
  .development(__DEVELOPMENT__)
  .alias(webpackConfig.resolve.alias)
  .server(webpackConfig.context, function() {
    require('../src/server')({
      webpackConfig: webpackConfig,
    });
  });
```

**Comment by halt-hammerzeit** [comment]
ok, I don't know why it might not work.
If you want you can share your project on github and I can try to run it.
Also do `console.log(webpackConfig.resolve.alias)` to be sure that it contains `{ 'theme': 'theme-a' }`

**Comment by x1a0** [comment]
> Also do console.log(webpackConfig.resolve.alias) to be sure that it contains { 'theme': 'theme-a' }

Yes that was what I did to verify and it's correct there.

I cannot share the whole project. Sorry.
Right now I am trying to dig into webpack-isomorphic-tools and require-hacker modules. Will update you here if I can find anything. Or I might create a show-case project to demo it.

As I said earlier, my current guess is:
in require-hacker, the extension resolver gets higher priority and run first so it does not use the aliasing resolver.

**Comment by halt-hammerzeit** [comment]
Ok, you can modify the source code and compile it with `npm run tryout` command.
It will output a `.tar.gz` archive which you can then `npm install ....tar.gz`.

I wrote `require-hacker` and it hooks at this line:
https://github.com/nodejs/node/blob/master/lib/module.js#L334

I wrote some tests for `alias` functionality and they are all passing:
https://github.com/halt-hammerzeit/webpack-isomorphic-tools/blob/master/test/index.js#L344-L376

Maybe the test case differs somehow from your case.

If you'll need explanation on some things I can answer the questions

**Comment by halt-hammerzeit** [comment]
Oh, I think I get it.
You don't require your scss as `require(...)`, right?
You import it instead in a scss file.
Which means that it takes place not in Node.js server code but instead inside Webpack plugin.

**Comment by x1a0** [comment]
No I import it in a `.jsx` file.

**Comment by halt-hammerzeit** [comment]
Post a line of code where you import it in a .jsx file

**Comment by x1a0** [comment]
const Styles = require('theme/sass/login.scss');

**Comment by halt-hammerzeit** [comment]
hmm, that should have worked.
`@import`s on the other hand aren't working yet.

**Comment by x1a0** [comment]
I am digging in `require-hacker` now. Added some debug log after this line:
https://github.com/halt-hammerzeit/require-hacker/blob/master/source/require%20hacker.js#L36

When the `request` is `theme/sass/login.scss`, `this.preceding_abstract_path_resolvers` is `[]`, so it goes to original Node.js loader I think.

**Comment by halt-hammerzeit** [comment]
`this.preceding_abstract_path_resolvers` should not be empty after `.alias()` method call because it creates a preceding hook

https://github.com/halt-hammerzeit/webpack-isomorphic-tools/blob/master/source/index.js#L119-L141

**Comment by halt-hammerzeit** [comment]
By the way I might have found the error.
Work in progress.

**Comment by x1a0** [comment]
OK. I feel there are different instances of `Require_hacker` been used...

**Comment by halt-hammerzeit** [comment]
Yes, one for plugin and one for server side.
I think I found the error: plugin funcitonality is absent and server side functionality has a bug.
I'm fixing it now.

**Comment by halt-hammerzeit** [comment]
Ok, I've made some changes, try this one out:
https://yadi.sk/d/m760Hu2WkH3XV

`npm install webpack-isomorphic-tools-2.2.9.tgz`

And post back your results.

Require-hacker is now 2.0, some api naming changes

**Comment by halt-hammerzeit** [comment]
Oh, and I forgot to tell: `{ alias: { ... } }` is now moved to the options instead of being called as a method because it's used both in plugin and in server instance

**Comment by x1a0** [comment]
No luck, same error. `this.preceding_abstract_path_resolvers` is still `[]` when i require `theme/sass/login.scss`.

**Comment by halt-hammerzeit** [comment]
Well, I've just tested it in my project with aliasing and my `preceding_abstract_path_resolvers` is not empty.

```

module.exports = 
{
	// debug: true,

	webpack_assets_file_path: './build/webpack-assets.json',

	alias:
	{
		'original_module_name': 'aliased_module_name'
	},

	assets:
	{
...
```

```
console.log('@@@@@@@@@@@@@@@@@@@@@@', global.webpack_isomorphic_tools.require_hacker.preceding_abstract_path_resolvers)
```

```
@@@@@@@@@@@@@@@@@@@@@@ [ { id: 'aliasing', resolve: [Function: resolve] } ]
```

**Comment by halt-hammerzeit** [comment]
Also, I've created a dummy module inside `node_modules` named `aliased_module_name`.
Then I added `alias` both in webpack config and in tools config.
Then I did `require('original_module_name')` and it throws an Error if I run this code outside the tools, and it resolves the module correctly when I run it with tools.
So in my case it seems working.

If you'll have any ideas you can post here.

**Comment by x1a0** [comment]
If I log `global.webpack_isomorphic_tools.require_hacker.preceding_abstract_path_resolvers`, it is *not* empty.
However if i console.log inside require-hacker file, `console.log(this.preceding_abstract_path_resolvers)` it is empty.

Could you register `.scss` ext in tools config then try something like `require('original_module_name/test.scss')`?

**Comment by x1a0** [comment]
In my case, the `tools` instance for plugin only has `aliasing` resolver while the instance for server-side only has `webpack-module` resolver.

**Comment by halt-hammerzeit** [comment]
The instance for the server side must also have `aliasing` resolver if `alias` option was supplied in the config.
If it's not there - then I don't know why.

**Comment by halt-hammerzeit** [comment]
By the way, `webpack-module` is only in plugin, it cannot be on the server.

**Comment by x1a0** [comment]
Sorry I was a bit lost in debugging...

    console.log(global.webpackIsomorphicTools.require_hacker.preceding_abstract_path_resolvers);

gives:

    [ { id: 'aliasing', resolve: [Function: resolve] } ]

I think I was wrong about those 2 in last message, should be the other way around.

**Comment by halt-hammerzeit** [comment]
So, `aliasing` hook is installed on the server.
Now, where you do this console.log, write an extra line with `require(original_module_name)` and see if it's aliased to aliased_module_name

**Comment by x1a0** [comment]
nope - it does not alias.

Here is my project: https://github.com/x1a0/showcase.
(need to use the tgz tools module you sent me: https://yadi.sk/d/m760Hu2WkH3XV)

This line is where the error comes:
https://github.com/x1a0/showcase/blob/master/src/app/views/App.jsx#L6

**Comment by halt-hammerzeit** [comment]
Oh, I see, so you're running both webpack and Node.js in the same process.
I get the error if I refresh the page after `webpack built`.
If I refresh before - then everything works.
Possibly webpack plugin somehow pollutes something.
I'll take a look.

**Comment by halt-hammerzeit** [comment]
One more idea: when webpack plugin does `.unmount()` for the `alias` hook - then it also unmounts the `alias` hook which was previously installed by tools.

**Comment by halt-hammerzeit** [comment]
So, the reason seems to be Require Hacker - each time you create a new instance, it overrides the system function.

```

	constructor(options)
	{
		// // take the passed in options
		// this.options = clone(options)

		// logging
		this.log = new Log('require-hook', { debug: options.debug }) // this.options.debug

		// instrument Module._findPath
		// https://github.com/nodejs/node/blob/master/lib/module.js#L335-L341
		Module._findPath = (...parameters) =>
		{
```

I'll modify it in such a way that it will override the system function only once, and remove that code from constructor.
Then I'll post here.

**Comment by halt-hammerzeit** [comment]
Ok, so `npm install webpack-isomorphic-tools@latest --save`
It's 2.2.10 and it works for me on your project ("Show case" on red)

**Comment by x1a0** [comment]
YEP! It works like a charm :+1: 

Thank you very much ;)

**Comment by x1a0** [comment]
`2.2.11` breaks it again (`2.2.10` works fine). Got:

```
AssertionError: path must be a string
   at Module.require (module.js:362:3)
   at Object.alias_hook (/home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/babel-transpiled-modules/common.js:265:22)
   at /home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/babel-transpiled-modules/index.js:152:19
   at Object.resolve (/home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/node_modules/require-hacker/babel-transpiled-modules/require hacker.js:91:18)
   at Function._module3.default._findPath (/home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/node_modules/require-hacker/babel-transpiled-modules/require hacker.js:382:27)
   at Function.Module._resolveFilename (module.js:332:25)
   at Function._module3.default._resolveFilename (/home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/node_modules/require-hacker/babel-transpiled-modules/require hacker.js:350:34)
   at Function.Module._load (module.js:284:25)
   at Module.require (module.js:363:17)
   at require (module.js:382:17)
```

**Comment by halt-hammerzeit** [comment]
Yes, I rewrote the `alias` functionality.
Post your debug logs when this exception is thrown.

**Comment by x1a0** [comment]
Sorry for late reply. The debug logs was in last message. Or did you mean something else?

**Comment by halt-hammerzeit** [comment]
@x1a0 I meant the rest of the console output.
You provided me with an exception stack trace.
Prior to this stack trace there are some debug messages.
Enable them by enabling `debug: true` flag.

**Comment by x1a0** [comment]
```
[webpack-isomorphic-tools] [debug] flushing require() caches
[webpack-isomorphic-tools] [debug] alias found for "theme/sass/login.scss", resolving to path "theme-a/sass/login.scss"
[webpack-isomorphic-tools] [debug]  global path for the aliased module is /home/zhangxiao/projects/theme-a/sass/login.scss
[webpack-isomorphic-tools] [debug]  normalized path for the aliased module is [object Object]
AssertionError: path must be a string
    at Module.require (module.js:362:3)
    at Object.alias_hook (/home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/babel-transpiled-modules/common.js:265:22)
    at /home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/babel-transpiled-modules/index.js:152:19
    at Object.resolve (/home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/node_modules/require-hacker/babel-transpiled-modules/require hacker.js:91:18)
    at Function._module3.default._findPath (/home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/node_modules/require-hacker/babel-transpiled-modules/require hacker.js:382:27)
    at Function.Module._resolveFilename (module.js:332:25)
    at Function._module3.default._resolveFilename (/home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/node_modules/require-hacker/babel-transpiled-modules/require hacker.js:350:34)
    at Function.Module._load (module.js:284:25)
    at Module.require (module.js:363:17)
    at require (module.js:382:17)
zzz

**Comment by halt-hammerzeit** [comment]
Oh, I see now - it returns object for some reason

**Comment by halt-hammerzeit** [comment]
This fix is quick - try the new version i've just released

**Comment by x1a0** [comment]
Still not working as `2.2.10`. Now I think it has some problem that has occurred before.

As you can see below, in assets file it uses relative path of `theme-a/sass/login.scss`, however in the debug log, or say when it resolved alias and tried to require the file, it used absolute path.

`../theme-a/sass/login.scss` and `/home/zhangxiao/projects/theme-a/sass/login.scss` are same file.


`webpack-assets.json`:

```
{
  "javascript": {
    "main": "/static/main-c481defbe3c6b8f12499.js"
  },
  "styles": {},
  "assets": {
    "./src/app/styles/main.scss": {
      "_style": "html, body, div, span, object, iframe, h1, h2, h3, h4, h5, h6, p, blockquote, pre, a, abbr, address, cite, code, del, dfn, em, img, ins, kbd, q, s, samp, small, strong, sub, sup, var, b, u, i, dl, dt, dd, ol, ul, li, fieldset, form, label, legend, table, caption, tbody, tfoot, thead, tr, th, td, article, aside, canvas, details, embed, figure, figcaption, footer, header, menu, nav, output, ruby, section, summary, time, mark, audio, video {\n  margin: 0;\n  padding: 0;\n  border: 0;\n  font-size: 100%;\n  font: inherit;\n  vertical-align: baseline;\n}\n\n/* HTML5 display-role reset for older browsers */\narticle, aside, details, figcaption, figure, footer, header, menu, nav, section {\n  display: block;\n}\n\nbody {\n  line-height: 1;\n}\n\nol, ul {\n  list-style: none;\n}\n\nblockquote, q {\n  quotes: none;\n}\n\nblockquote:before, blockquote:after {\n  content: '';\n  content: none;\n}\n\nq:before, q:after {\n  content: '';\n  content: none;\n}\n\ntable {\n  border-collapse: collapse;\n  border-spacing: 0;\n}\n"
    },
    "../theme-a/sass/login.scss": {
      "loginForm": "loginForm___faZqf",
      "_style": ".loginForm___faZqf {\n  background-color: #0000ff;\n}\n"
    }
  }
}
```

Debug log:

```
[webpack-isomorphic-tools] [debug] flushing require() caches
[webpack-isomorphic-tools] [debug] alias found for "theme/sass/login.scss", resolving to path "theme-a/sass/login.scss"
[webpack-isomorphic-tools] [debug]  global path for the aliased module is /home/zhangxiao/projects/theme-a/sass/login.scss
[webpack-isomorphic-tools] [debug]  normalized path for the aliased module is /home/zhangxiao/projects/theme-a/sass/login.scss
[require-hook] [debug] Loading source code for /home/zhangxiao/projects/theme-a/sass/login.scss
[webpack-isomorphic-tools] [debug] require() called for /home/zhangxiao/projects/theme-a/sass/login.scss
[webpack-isomorphic-tools] [debug]  requiring /home/zhangxiao/projects/theme-a/sass/login.scss
[webpack-isomorphic-tools] [error] asset not found: /home/zhangxiao/projects/theme-a/sass/login.scss
[webpack-isomorphic-tools] [debug]  the path was found
[require-hook] [debug] Loading source code for theme/sass/login.scss.aliasing
```

**Comment by halt-hammerzeit** [comment]
ok, I'll make it always relative

**Comment by halt-hammerzeit** [comment]
try the newest version
2.2.13

**Comment by x1a0** [comment]
Nope... now it gives:

```
[webpack-isomorphic-tools] [debug] flushing require() caches
[webpack-isomorphic-tools] [debug] alias found for "theme/sass/login.scss", resolving to path "theme-a/sass/login.scss"
[webpack-isomorphic-tools] [debug]  global path for the aliased module is /home/zhangxiao/projects/theme-a/sass/login.scss
[webpack-isomorphic-tools] [debug]  normalized path for the aliased module is ../theme-a/sass/login.scss
Error: Cannot find module '../theme-a/sass/login.scss'
    at Function.Module._resolveFilename (module.js:334:15)
    at Function._module3.default._resolveFilename (/home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/node_modules/require-hacker/babel-transpiled-modules/require hacker.js:350:34)
    at Function.Module._load (module.js:284:25)
    at Module.require (module.js:363:17)
    at Object.alias_hook (/home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/babel-transpiled-modules/common.js:265:22)
    at /home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/babel-transpiled-modules/index.js:152:19
    at Object.resolve (/home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/node_modules/require-hacker/babel-transpiled-modules/require hacker.js:91:18)
    at Function._module3.default._findPath (/home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/node_modules/require-hacker/babel-transpiled-modules/require hacker.js:382:27)
    at Function.Module._resolveFilename (module.js:332:25)
    at Function._module3.default._resolveFilename (/home/zhangxiao/projects/cth-poc/node_modules/webpack-isomorphic-tools/node_modules/require-hacker/babel-transpiled-modules/require hacker.js:350:34)
```

assets file is the same:

```
{
  "javascript": {
    "main": "/static/main-c481defbe3c6b8f12499.js"
  },
  "styles": {},
  "assets": {
    "./src/app/styles/main.scss": {
      "_style": "html, body, div, span, object, iframe, h1, h2, h3, h4, h5, h6, p, blockquote, pre, a, abbr, address, cite, code, del, dfn, em, img, ins, kbd, q, s, samp, small, strong, sub, sup, var, b, u, i, dl, dt, dd, ol, ul, li, fieldset, form, label, legend, table, caption, tbody, tfoot, thead, tr, th, td, article, aside, canvas, details, embed, figure, figcaption, footer, header, menu, nav, output, ruby, section, summary, time, mark, audio, video {\n  margin: 0;\n  padding: 0;\n  border: 0;\n  font-size: 100%;\n  font: inherit;\n  vertical-align: baseline;\n}\n\n/* HTML5 display-role reset for older browsers */\narticle, aside, details, figcaption, figure, footer, header, menu, nav, section {\n  display: block;\n}\n\nbody {\n  line-height: 1;\n}\n\nol, ul {\n  list-style: none;\n}\n\nblockquote, q {\n  quotes: none;\n}\n\nblockquote:before, blockquote:after {\n  content: '';\n  content: none;\n}\n\nq:before, q:after {\n  content: '';\n  content: none;\n}\n\ntable {\n  border-collapse: collapse;\n  border-spacing: 0;\n}\n"
    },
    "../theme-a/sass/login.scss": {
      "loginForm": "loginForm___faZqf",
      "_style": ".loginForm___faZqf {\n  background-color: #0000ff;\n}\n"
    }
  }
}
```

**Comment by halt-hammerzeit** [comment]
Ok, try v2.2.14

**Comment by x1a0** [comment]
Everything seems fine now. Thanks!

**Comment by halt-hammerzeit** [comment]
@x1a0 Hello. Today I released a patched version to address some aliasing issues
https://github.com/halt-hammerzeit/require-hacker/issues/1#issuecomment-161971347
Supposedly it won't break existing code.
You are informed now.

**Comment by halt-hammerzeit** [comment]
FYI
I'm posting this in every issue and PR to notify whoever may be interested:
today I've released an alternative helper library called [`universal-webpack`](https://github.com/halt-hammerzeit/universal-webpack).
It takes a different approach than `webpack-ismorphic-tools` and instead of hacking Node.js `require()` calls it just compiles all code with `target: 'node'` webpack configuration option.
As a result, all Webpack plugins and features are supported.
If you think you might need that here's an example project:
https://github.com/halt-hammerzeit/webpack-react-redux-isomorphic-render-example
